### PR TITLE
atks start

### PR DESCRIPTION
**Pull request by Kami78kami**
APGD added, PGD and BIM as well as APGD all changed to fixed targeted mode. Also in AttackDS there were some changes to store target labels for further evaluations.

### CODE REVIEW

File: peepholelib/adv_atk/APGD_old.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
+# torch stuff
```
Comment (lorenzoCapelli) [inline peepholelib/adv_atk/APGD_old.py]: What is the diference between APGD.py and APGD_old.py

File: peepholelib/adv_atk/APGD.py [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+                ).detach() * self.normalize(t)
+        elif self.norm == 'L2':
+            t = torch.randn(x.shape).to(self.device).detach()
+            x_adv = x + self.eps * torch.ones_like(x
+                ).detach() * self.normalize(t)
+        elif self.norm == 'L1':
+            t = torch.randn(x.shape).to(self.device).detach()
+            delta = L1_projection(x, t, self.eps)
+            x_adv = x + t + delta
+
```
Comment (lorenzoCapelli) [inline peepholelib/adv_atk/APGD.py]: remove empty lines

File: peepholelib/adv_atk/APGD.py [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+                      x_adv[fl_redtopk] = x_best[fl_redtopk].clone()
+                      grad[fl_redtopk] = grad_best[fl_redtopk].clone()
+                  
+                  counter3 = 0
+                  #k = max(k - self.size_decr, self.n_iter_min)
+
+        #
+        
+        return (x_best, acc, loss_best, x_best_adv)
+class perturbnew(APGDTnew): # APGDAttack_targeted was OG
```
Comment (lorenzoCapelli) [inline peepholelib/adv_atk/APGD.py]: A class with a single method doesn't make sense to make `perturb` as a method of `APGDTnew`?

File: peepholelib/adv_atk/attacksDS.py [hunk 4]
```diff
@@ -83,6 +83,7 @@ def apply_attacks(self, **kwargs):
                         )
 
                 self._dss[tdsk]['attack_success'] = MMT.empty(shape=torch.Size((n_samples,)))
+                self._dss[tdsk]['y_target'] = MMT.empty(shape=torch.Size((n_samples,)), dtype=torch.long) # by Kami
```
Comment (lorenzoCapelli) [inline peepholelib/adv_atk/attacksDS.py:86]: remove comment with name

File: peepholelib/adv_atk/attacksDS.py [hunk 5]
```diff
@@ -123,19 +124,44 @@ def apply_attacks(self, **kwargs):
...
+                        y_target = y_target.detach().cpu().long()
+                    # added by Kami
                     
                     with torch.no_grad():
                         y_pred_ori = atk.model(ori_images.to(atk.model.device))
                         y_pred_atk = atk.model(atk_images.to(atk.model.device))
                     pred_labels_ori = y_pred_ori.argmax(axis = 1)
                     pred_labels_atk = y_pred_atk.argmax(axis = 1)
 
+<<<<<<< HEAD
```
Comment (lorenzoCapelli) [inline peepholelib/adv_atk/attacksDS.py]: Something went wrong with the rebase

File: peepholelib/adv_atk/attacksDS_old.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
+# general python stuff
```
Comment (lorenzoCapelli) [inline peepholelib/adv_atk/attacksDS_old.py]: If it is deprecated eliminate it

File: peepholelib/plots/useless_plot.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
+from pathlib import Path
```
Comment (lorenzoCapelli) [inline peepholelib/plots/useless_plot.py]: if it is useless eliminate it

File: peepholelib/datasets/parsedDataset.py [hunk 8]
```diff
@@ -199,6 +199,7 @@ def parse_ds(cls, **kwargs):
                             for key in out_ktc:
                                 data_t[key] = _res[key]
 
+        cls_inst._dss = None
```
Comment (lorenzoCapelli) [inline peepholelib/datasets/parsedDataset.py]: why are you imposing the attribute euqal to None?

File: peepholelib/datasets/parsedDataset_og.py [hunk 9]
```diff
@@ -0,0 +1,254 @@
+# General python stuff
```
Comment (lorenzoCapelli) [inline peepholelib/datasets/parsedDataset_og.py]: Why a new version of the the parsed dataset has been created?

File: peepholelib/adv_atk/PGD.py [hunk 10]
```diff
@@ -31,27 +37,48 @@ def __init__(self, **kwargs):
...
+                return y_t
+
+            self.atk.set_mode_targeted_by_function(fixed_target_fn, quiet=False)
+
+        elif self.mode == "custom":
+            def custom_target_fn(inputs, labels):
+                y_t = self.custom_target_labels.to(labels.device).long()
+                return y_t
+            self.atk.set_mode_targeted_by_function(custom_target_fn, quiet=False)
+
```
Comment (lorenzoCapelli) [inline peepholelib/adv_atk/PGD.py:48]: missing the return

File: peepholelib/adv_atk/CW.py [hunk 11]
```diff
@@ -615,6 +615,7 @@ def __init__(self, **kwargs):
         self.max_steps = kwargs.get('max_steps', 1000)
         self.optimizer_lr = kwargs.get('optimizer_lr', 1e-2)
         self.mode = kwargs.get('mode', 'random')
+        self.n_classes = kwargs.get("n_classes", 100)   # for CIFAR-100 obv
```
Comment (lorenzoCapelli) [inline peepholelib/adv_atk/CW.py]: i would impose to the user to choose the number of classes just to be sure. Eliminate the default.

File: peepholelib/adv_atk/PGD.py [hunk 12]
```diff
@@ -31,27 +37,48 @@ def __init__(self, **kwargs):
...
+                if same.any():
+                    with torch.no_grad():
+                        num_classes = self.atk.get_output_with_eval_nograd(inputs).shape[1]
+                    y_t[same] = (y_t[same] + 1) % num_classes
+                return y_t
+
+            self.atk.set_mode_targeted_by_function(fixed_target_fn, quiet=False)
+
+        elif self.mode == "custom":
+            def custom_target_fn(inputs, labels):
```
Comment (lorenzoCapelli) [inline peepholelib/adv_atk/PGD.py]: missing what is happening here

File: peepholelib/adv_atk/BIM.py [hunk 13]
```diff
@@ -29,28 +42,40 @@ def __init__(self, **kwargs):
...
+
+        elif self.mode == "least-likely":
             self.atk.set_mode_targeted_least_likely(kth_min=1, quiet=False)
-            self.atk.get_least_likely_label
 
-        return
+        elif self.mode == "fixed":
+            tc = int(self.target_class)
+
+            def fixed_target_fn(inputs, labels):
```
Comment (lorenzoCapelli) [inline peepholelib/adv_atk/BIM.py]: this function is used only for BIM and PGD however i would place it as a attack_base method so you do not have to repeat it in different classes

File: peepholelib/adv_atk/BIM.py [hunk 14]
```diff
@@ -1,9 +1,22 @@
 # torch stuff
+import torch
 import torchattacks
 
 # our stuff
 from .attack_base import AttackBase
 
+
+class BIMStoreTarget(torchattacks.BIM):
+    def get_target_label(self, inputs, labels=None):
```
Comment (lorenzoCapelli) [inline peepholelib/adv_atk/BIM.py]: When is this method used?

File: peepholelib/adv_atk/BIM.py [hunk 15]
```diff
@@ -1,9 +1,22 @@
 # torch stuff
+import torch
 import torchattacks
 
 # our stuff
 from .attack_base import AttackBase
 
+
+class BIMStoreTarget(torchattacks.BIM):
```
Comment (lsrosa) [inline peepholelib/adv_atk/BIM.py]: It feels unnecessary to make another class to wrap BIM just to store the target label, when `myBIM` is already an wrapper to `torchattacks.BIM`, you could move this inside `myBIM`. Or maybe in attkBase

File: peepholelib/adv_atk/APGD.py [hunk 16]
```diff
@@ -0,0 +1,416 @@
...
+
+class APGDTnew(APGDAttack_targeted):
+    def attack_single_run(self, x, y, x_init=None):
+        if len(x.shape) < self.ndims:
+            x = x.unsqueeze(0)
+            y = y.unsqueeze(0)
+
+        if self.norm == 'Linf':
+            t = 2 * torch.rand(x.shape).to(self.device).detach() - 1
+            x_adv = x + self.eps * torch.ones_like(x
```
Comment (lsrosa) [inline peepholelib/adv_atk/APGD.py]: Can you remove these linebreaks in the middle of expresions?

File: peepholelib/adv_atk/APGD.py [hunk 17]
```diff
@@ -0,0 +1,416 @@
...
+from autoattack.autopgd_base import APGDAttack, APGDAttack_targeted
+from .attack_base import AttackBase
+
+import torch.nn as nn
+import torch.nn.functional as F
+from autoattack.other_utils import L0_norm, L1_norm, L2_norm
+from autoattack.checks import check_zero_gradients
+from autoattack.autopgd_base import L1_projection
+
+class APGDTnew(APGDAttack_targeted):
```
Comment (lsrosa) [inline peepholelib/adv_atk/APGD.py]: `new` is not a good name, what is the difference between this class and the `myAPGD`? probably `targeted`

File: peepholelib/adv_atk/attacksDS.py
Comment (lsrosa) [inline peepholelib/adv_atk/attacksDS.py:19]: This class will be removed on #151 , you should probably prepare for the merge

File: peepholelib/adv_atk/attacksDS.py [hunk 18]
```diff
@@ -123,19 +124,35 @@ def apply_attacks(self, **kwargs):
                             images = ori_images,
                             labels = labels 
                             )
+                    # added by Kami
+                    y_target = getattr(atk.atk, "y_target", None)
```
Comment (lsrosa) [inline peepholelib/adv_atk/attacksDS.py]: please avoid using `getattr`, you probably should init `y_target = None` and access it directly.

File: peepholelib/adv_atk/BIM.py [hunk 19]
```diff
@@ -29,28 +42,41 @@ def __init__(self, **kwargs):
...
+        # self.target_class = kwargs.get("target_class", 5)
+
+        self.atk = BIMStoreTarget(
+            model=self.model._model,
+            eps=self.eps,
+            alpha=self.alpha,
+            steps=self.steps,
+        )
+
+        # if self.mode == "random":
```
Comment (lsrosa) [inline peepholelib/adv_atk/BIM.py]: clean commnts

File: peepholelib/plots/AllSeeing.py [hunk 20]
```diff
@@ -0,0 +1,205 @@
+from pathlib import Path
+import torch
+import matplotlib.pyplot as plt
+from tensordict import PersistentTensorDict
+
+
```
Comment (lsrosa) [inline peepholelib/plots/AllSeeing.py:5]: `AllSeeing` is a weird name, can we use a proper one?

File: peepholelib/plots/attack_compare.py [hunk 21]
```diff
@@ -0,0 +1,193 @@
...
+      }
+    """
+
+    def __init__(self, base_path, dataset="CIFAR100", threat_models=None):
+        self.base = Path(base_path)
+        self.dataset = dataset
+        self.threat_models = threat_models or {}
+
+    def _load(self, name):
+        return PersistentTensorDict.from_h5(self.base / name, mode="r")
```
Comment (lsrosa) [inline peepholelib/plots/attack_compare.py:23]: As per standard, you should pass the datasets already loaded. Note that you are missing the `close()`s, which implies you are leaving the files open. the loaded dataset manages these things for you.

**Review by lsrosa — CHANGES_REQUESTED** [review]
The code is full of `if`s, it is a bit hard to read. We should find a better way to set these different options.

There are many unnecessary data casts between GPU and CPU, we also need to clean that out

File: peepholelib/adv_atk/BIM.py (hunk 13 above)
Comment (Kami78kami) [reply peepholelib/adv_atk/BIM.py]: Done, moved to attack_base

File: peepholelib/adv_atk/BIM.py (hunk 14 above)
Comment (Kami78kami) [reply peepholelib/adv_atk/BIM.py]: This one is used to store the target labels as the original torchattacks do not store or return them (it is just used internally).

File: peepholelib/plots/AllSeeing.py (hunk 20 above)
Comment (Kami78kami) [reply peepholelib/plots/AllSeeing.py:5]: Sorry I didnt mean to push this one, it was for my own interpretation

File: peepholelib/plots/attack_compare.py (hunk 21 above)
Comment (Kami78kami) [reply peepholelib/plots/attack_compare.py:23]: same for this, I didnt mean to push this one, it was for my own interpretation. Sorry for inconvenience.

File: peepholelib/adv_atk/BIM.py (hunk 15 above)
Comment (Kami78kami) [reply peepholelib/adv_atk/BIM.py]: Done, I moved it to attackBase for now (I haven't pushed yet to do the other comments)

File: peepholelib/adv_atk/BIM.py (hunk 14 above)
Comment (Kami78kami) [reply peepholelib/adv_atk/BIM.py]: Now moved to attackBase

File: peepholelib/adv_atk/APGD.py (hunk 16 above)
Comment (Kami78kami) [reply peepholelib/adv_atk/APGD.py]: Done. It was because I copy pasted the code from AutoAttack original code, and just changed and added the utilities I wanted and left the rest unchanged. Now there shouldn't be any more of those line breaks, i cleaned them all.

File: peepholelib/adv_atk/APGD.py (hunk 17 above)
Comment (Kami78kami) [reply peepholelib/adv_atk/APGD.py]: Yes I wasn't paying attention to names at the time. It is targeted which inherits from APGD targeted version of autoattack and instead of automatic target selection, selects the targets provided (fixed, least likely and random)

Anyways I will change the name
